cesql/spec.md

```diff
@@ -355,26 +396,26 @@ left operand of the OR operation evalues to `true`, the right operand MUST NOT b
 
 When the argument types of an operator/function invocation don't match the signature of the operator/function being invoked, the CESQL engine MUST try to perform an implicit cast.
 
-This section defines an **ambiguous** operator/function as an operator/function that is overloaded with another
-operator/function definition with same symbol/name and arity but different parameter types.
+This section defines an **ambiguous** operator as an operator that is overloaded with another
+operator definition with same symbol/name and arity but different parameter types. Note: a function
+can not be ambiguous as it is not allowed for two functions to have the same arity and name.
 
 A CESQL engine MUST apply the following implicit casting rules in order:
 
 1. If the operator/function is unary (argument `x`):
    1. If it's not ambiguous, cast `x` to the parameter type.
-   1. If it's ambiguous, raise an error and the cast result is undefined.
+   1. If it's ambiguous, raise a _CastError_ and the cast result is `false`.
 1. If the operator is binary (left operand `x` and right operand `y`):
    1. If it's not ambiguous, cast `x` and `y` to the corresponding parameter types.
    1. If it's ambiguous, use the `y` type to search, in the set of ambiguous operators, every definition of the operator
       using the `y` type as the right parameter type:
       1. If such operator definition exists and is unique, cast `x` to the type of the left parameter
-      1. Otherwise, raise an error and the cast results are undefined
+      1. Otherwise, raise a _CastError_ and the result is `false`
 1. If the function is n-ary with `n > 1`:
-   1. If it's not ambiguous, cast all the arguments to the corresponding parameter types.
-   1. If it's ambiguous, raise an error and the cast results are undefined.
+   1. Cast all the arguments to the corresponding parameter types.
 1. If the operator is n-ary with `n > 2`:
```

Comment:
I'm sure there a reason, but why "n>2" and not "n>0" ?

And I guess on the previous one, why "n > 1" and not "n > 0" ?

Reply:
For "n>2" instead of "n>0" this is because we already handled the "n=1" and "n=2" earlie with the unary and binary operator/function sections.

As for why we only handle binary operator resolution and not binary function resolution (the "n>1" vs "n>2"), I am not sure, this comes from before my time working on the spec. I'm inclined to believe there was a reason for it though